File: code/data_structures/src/tree/multiway_tree/union_find/union_find.cpp
#include <iostream>
#include <vector>

/* Part of Cosmos by OpenGenus Foundation */

// Union-find stores a graph, and allows two operations in amortized constant time:
// * Add a new edge between two vertices.
// * Check if two vertices belong to same component.
class UnionFind {
    std::vector<size_t> parent;
    std::vector<size_t> rank;
    size_t root(size_t node)
    {
        if (parent[node] == node)
            return node;
        else
            return parent[node] = getParent(parent[node]);
    }

    size_t getParent(size_t node)
    {
        return parent[node];
    }

public:
// Make a graph with `size` vertices and no edges.
    UnionFind(size_t size)
    {
        parent.resize(size);
        rank.resize(size);
        for (size_t i = 0; i < size; i++)
        {
            parent[i] = i;
            rank[i] = 1;
        }
    }

// Connect vertices `a` and `b`.
    void Union(size_t a, size_t b)
    {
        a = root(a);
        b = root(b);
        if (a == b)
            return;
        if (rank[a] < rank[b])
            parent[a] = b;
        else if (rank[a] > rank[b])
            parent[b] = a;
        else
        {
            parent[a] = b;
            rank[b] += 1;
        }
    }

// Check if vertices `a` and `b` are in the same component.
    bool Connected(size_t a, size_t b)
    {
        return root(a) == root(b);
    }
};

int main()
{
    UnionFind unionFind(10);

    unionFind.Union(3, 4);
    unionFind.Union(3, 8);
    unionFind.Union(0, 8);
    unionFind.Union(1, 3);
    unionFind.Union(7, 9);
    unionFind.Union(5, 9);

    // Now the components are:
    // 0 1 3 4 8
    // 5 7 9
    // 2
    // 6

    for (size_t i = 0; i < 10; i++)
        for (size_t j = i + 1; j < 10; j++)
            if (unionFind.Connected(i, j))
                std::cout << i << " and " << j << " are in the same component" << std::endl;

    return 0;
}

File: code/data_structures/src/tree/multiway_tree/union_find/union_find.js
/* Part of Cosmos by OpenGenus Foundation */

class UnionFind {
  constructor() {
    this.parents = [];
    this.ranks = [];
  }

  find(v) {
    if (this.parents[v] === v) {
      return v;
    }
    this.parents[v] = this.find(this.parents[v]);
    return this.parents[v];
  }

  union(a, b) {
    if (this.parents[a] === undefined) {
      this.parents[a] = a;
      this.ranks[a] = 0;
    }

    if (this.parents[b] === undefined) {
      this.parents[b] = b;
      this.ranks[b] = 0;
    }

    a = this.find(a);
    b = this.find(b);

    if (a !== b) {
      if (this.ranks[a] < this.ranks[b]) {
        [a, b] = [b, a];
      }

      if (this.ranks[a] === this.ranks[b]) {
        this.ranks[a]++;
      }

      this.parents[b] = a;
    }
  }

  connected(a, b) {
    return this.find(a) === this.find(b);
  }
}

const uf = new UnionFind();
uf.union(1, 3);
uf.union(1, 4);
uf.union(4, 2);
uf.union(5, 6);
console.log(uf.find(3));
console.log(uf.find(4));
console.log(uf.find(2));
console.log(uf.find(5));
